core/src/main/java/com/linecorp/armeria/client/Proxy.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.net.InetSocketAddress;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.client.ProxyBuilder.ConnectProxyBuilder;
+import com.linecorp.armeria.client.ProxyBuilder.Socks4ProxyBuilder;
+import com.linecorp.armeria.client.ProxyBuilder.Socks5ProxyBuilder;
+
+/**
+ * Contains base configuration for proxy related settings used in {@link ClientFactory}.
+ */
+public class Proxy {
```

Comment:
- `final`
- Probably `ProxyConfig` is better?
- Please implement `toString()` for this class and its subclasses.

Reply:
Couldn't use `final` since other classes are inheriting this class